src/tools/cephfs/DataScan.cc

```diff
@@ -0,0 +1,1267 @@
+    inodeno_t inono, int mode, int64_t data_pool_id)
+{
+  // Compose
+  object_t oid = InodeStore::get_object_name(inono, frag_t(), ".inode");
+  InodeStore inode;
+  inode.inode.ino = inono;
+  inode.inode.version = 1;
+  inode.inode.xattr_version = 1;
+  inode.inode.mode = 0500 | mode;
+  // Fake size to 1, so that the directory doesn't appear to be empty
+  // (we won't actually give the *correct* size here though)
+  inode.inode.size = 1;
+  inode.inode.dirstat.nfiles = 1;
+
+  inode.inode.ctime = 
+    inode.inode.mtime = ceph_clock_now(g_ceph_context);
+  inode.inode.nlink = 1;
+  inode.inode.truncate_size = -1ull;
+  inode.inode.truncate_seq = 1;
+
+  // Force layout to default: should we let users override this so that
+  // they don't have to mount the filesystem to correct it?
+  inode.inode.layout = g_default_file_layout;
+  inode.inode.layout.fl_pg_pool = data_pool_id;
+
+  // Serialize
+  bufferlist inode_bl;
+  ::encode(std::string(CEPH_FS_ONDISK_MAGIC), inode_bl);
+  inode.encode(inode_bl);
+
+  // Write
+  int r = metadata_io.write_full(oid.name, inode_bl);
+  if (r != 0) {
+    derr << "Error writing '" << oid.name << "': " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  return r;
+}
+
+int MetadataDriver::root_exists(inodeno_t ino, bool *result)
+{
+  object_t oid = InodeStore::get_object_name(ino, frag_t(), ".inode");
+  uint64_t size;
+  time_t mtime;
+  int r = metadata_io.stat(oid.name, &size, &mtime);
+  if (r == -ENOENT) {
+    *result = false;
+    return 0;
+  } else if (r < 0) {
+    return r;
+  }
+
+  *result = true;
+  return 0;
+}
+
+int MetadataDriver::init_roots(int64_t data_pool_id)
+{
+  int r = 0;
+  r = inject_unlinked_inode(MDS_INO_ROOT, S_IFDIR|0755, data_pool_id);
+  if (r != 0) {
+    return r;
+  }
+  r = inject_unlinked_inode(MDS_INO_MDSDIR(0), S_IFDIR, data_pool_id);
+  if (r != 0) {
+    return r;
+  }
+
+  return 0;
+}
+
+int MetadataDriver::check_roots(bool *result)
+{
+  int r;
+  r = root_exists(MDS_INO_ROOT, result);
+  if (r != 0) {
+    return r;
+  }
+  if (!*result) {
+    return 0;
+  }
+
+  r = root_exists(MDS_INO_MDSDIR(0), result);
+  if (r != 0) {
+    return r;
+  }
+  if (!*result) {
+    return 0;
+  }
+
+  return 0;
+}
+
+/**
+ * Stages:
+ * PARALLEL
+ *  1. Size and mtime recovery: scan ALL objects, and update 0th
+ *   objects with max size and max mtime seen.
+ * PARALLEL
+ *  2. Inode recovery: scan ONLY 0th objects, and inject metadata
+ *   into dirfrag OMAPs, creating blank dirfrags as needed.  No stats
+ *   or rstats at this stage.  Inodes without backtraces go into
+ *   lost+found
+ * SERIAL
+ *  3. Dirfrag statistics: depth first traverse into metadata tree,
+ *    rebuilding dir sizes.
+ *    XXX hmm, *or* we could use an object class to accumulate
+ *    the number of insertions we've done on any given frag, and
+ *    then do a pass where we just inject that accumulated number
+ *    of insertions into the directory's size and the dirfrags stats?
```

Comment:
I haven't been all the way through even the first commit, but yes, we should do that. It can happen in the same stage as inode recovery and would have basically the same cost (what with disk hits and ops-to-process being the dominant factors).